### DISCUSSION TITLE
Shared Module Not Available for Eager Consumption

### DISCUSSION THREAD

**Discussion by zgevorg** [webpack/webpack#17890] [Q&A] [answered]
### Context

I have a React application initially created with Create React App (CRA), where I modify the Webpack configuration using CRACO. Additionally, I have developed a simple Microfrontend (MF) application that directly utilizes Webpack for its setup.

### Issue

I'm facing an issue related to Module Federation in this setup. The problem occurs when I try to share modules between my CRA application (configured with CRACO) and the separate MF application. The specific challenge is an error indicating difficulties with the eager consumption of a shared module.

The error message I encounter is:

`Uncaught Error: Shared module is not available for eager consumption: webpack/sharing/consume/default/react/react`
<img width="783" alt="Screenshot 2023-12-15 at 18 45 54" src="https://github.com/webpack/webpack/assets/16259486/6dd9a7c4-5e5a-4620-a985-01feb45a16b5">

### Troubleshooting Efforts

- I have extensively searched through many discussions, forums, and Webpack's official documentation, trying out various suggested solutions, but none have resolved the issue.
- I've carefully configured the shared dependencies in both the CRA and MF applications.
- I've confirmed the successful loading and response of remoteEntry.js from the MF application.
- I've experimented with various configurations, including removing the eager flag, in an attempt to solve this issue.

### Environment

**Container app**

- React version: 18.2.0
- CRA version: 5.0.1
- CRACO version: 7.1.0

**MF app**
- React version: 18.2.0
- Webpack version: 5.89.0

### Project Structure and Code Samples

**Container app**

_craco.config.js_ 

<img width="947" alt="Screenshot 2023-12-15 at 18 47 44" src="https://github.com/webpack/webpack/assets/16259486/00fc6236-edc5-40e5-a8d7-b2ace7ac75d2">

_App.jsx_

<img width="1027" alt="Screenshot 2023-12-15 at 18 49 34" src="https://github.com/webpack/webpack/assets/16259486/88ad080a-be58-480a-bed2-d0db3c6f5329">

**MF app**

_structure_

<img width="573" alt="Screenshot 2023-12-15 at 18 50 11" src="https://github.com/webpack/webpack/assets/16259486/c21fc0e8-fede-4c11-a821-b79f97af486f">

_webpack.config.js_

<img width="1029" alt="Screenshot 2023-12-15 at 18 51 49" src="https://github.com/webpack/webpack/assets/16259486/33c05595-e18b-4aed-a496-05515e91d4db">
<img width="1011" alt="Screenshot 2023-12-15 at 18 52 21" src="https://github.com/webpack/webpack/assets/16259486/6d170860-0233-4133-bda1-ff8fc232a1ee">

_index.js_
<img width="1032" alt="Screenshot 2023-12-15 at 18 52 42" src="https://github.com/webpack/webpack/assets/16259486/2d50f128-3824-4a43-897c-2e6f795be998">

_bootstrap.js_
<img width="1032" alt="Screenshot 2023-12-15 at 18 52 56" src="https://github.com/webpack/webpack/assets/16259486/e664f4d1-c67b-4871-91fa-9996d36adf27">

_App.jsx_
<img width="1032" alt="Screenshot 2023-12-15 at 18 53 19" src="https://github.com/webpack/webpack/assets/16259486/ca35e64d-796f-45ae-a01d-5e89386b2d68">

Also I would like to share remoteEntry.js response
<img width="1117" alt="Screenshot 2023-12-15 at 18 55 04" src="https://github.com/webpack/webpack/assets/16259486/24977f28-dad5-4e39-a3f1-d16f9d2995d5">

**Comment by alexander-akait** [comment]
Please avoid using screenshots, they are unreadable and hard to undestand, will be great if you will create a reproducible test repo and show your problem, thank you

**Reply by zgevorg** [reply]
Hey @alexander-akait ,

Thanks for the response!

I've created a reproducible test [repo](https://github.com/zgevorg/mf-issue)

**Reply by zgevorg** [reply]
Hey @alexander-akait , 
did you have a chance/time to take a look?

**Reply by alexander-akait** [reply]
Please read https://webpack.js.org/concepts/module-federation/#troubleshooting, there is a solution

**Reply by zgevorg** [reply]
As I mentioned in the **Troubleshooting Efforts** part I've checked and tried different solutions including this one.

I've added it in my test [repo](https://github.com/zgevorg/mf-issue), but still receiving the same error

**Comment by karlavetyan** [comment]
Same problem :(

**Reply by azat2** [reply]
Same ++++

**Comment by Tigran2222** [comment]
++

**Comment by danielm-infotrack** [comment]
Same problem with same code as above. @zgevorg did you find any resolution?

**Reply by zgevorg** [reply]
still not. 

I was asked by @alexander-akait to create a reproducible repo with the problem. So I created it and looking forward to response 🙌

**Reply by danielm-infotrack** [reply, marked as answer]
Hey @zgevorg i checked out your repo and i can see the issue is the container app isnt equipped with the bootstrap
This PR fixes it:

https://github.com/zgevorg/mf-issue/pull/1

![image](https://github.com/webpack/webpack/assets/69128322/e625bd4f-05ac-4488-9fe0-4b2acfee090f)

![image](https://github.com/webpack/webpack/assets/69128322/577d3e5d-d326-4834-8523-1967ce441e69)

**Reply by zgevorg** [reply]
Thanks for the response!  It worked

**Reply by JunboJ** [reply]
This doc says this method has limitations and drawbacks, where can I learn more about the limitations and drawbacks?
![image](https://github.com/webpack/webpack/assets/48075577/5f630829-9b92-4abc-9853-18a3f1e52a7a)
https://webpack.js.org/concepts/module-federation/#uncaught-error-shared-module-is-not-available-for-eager-consumption
(not sure if it's talking about the bootstrap method or the eager option method, sorry for my bad english).
